File: C++/GPT-4/Queue/Easy/Runtime and Memory/2073. Time Needed to Buy Tickets/2.cpp
#include <queue>
#include <vector>

class Solution {
public:
    int timeRequiredToBuy(std::vector<int>& tickets, int k) {
        using pii = std::pair<int, int>; // pair of int
        auto cmp = [](const pii& left, const pii& right) { // custom compare function
            if (left.first != right.first) return left.first < right.first;
            return left.second > right.second;
        };
        std::priority_queue<pii, std::vector<pii>, decltype(cmp)> pq(cmp);

        for (int i = 0; i < tickets.size(); ++i) {
            pq.push({tickets[i], i});
        }

        int time = 0;
        while (!pq.empty()) {
            pii person = pq.top();
            pq.pop();
            time++;
            if (person.second == k && person.first == 1) {
                break;
            }
            if (person.first > 1) {
                pq.push({person.first - 1, person.second});
            } else if (person.second == k) {
                break;
            }
        }
        return time;
    }
};

File: Java/GPT-4/Queue/Easy/Runtime and Memory/2073. Time Needed to Buy Tickets/2.java
import java.util.*;

class Solution {
    class Pair {
        int idx, tickets;
        public Pair(int idx, int tickets) {
            this.idx = idx;
            this.tickets = tickets;
        }
    }

    public int timeRequiredToBuy(int[] tickets, int k) {
        Deque<Pair> queue = new LinkedList<>();
        
        for (int i = 0; i < tickets.length; i++) {
            queue.offer(new Pair(i, tickets[i]));
        }
        
        int time = 0;
        while (!queue.isEmpty()) {
            Pair current = queue.poll();
            current.tickets--;
            time++;
            
            if (current.idx == k && current.tickets == 0) {
                return time;
            }
            
            if (current.tickets > 0) {
                queue.offer(current);
            }
        }
        return -1; // Shouldn't reach here as we are guaranteed to be able to sell all tickets
    }
}

